docs/ci.md

```diff
@@ -54,3 +54,201 @@ Received 0 of 614674276 (0.0%), 0.0 MBs/sec
+for more ways of specifying tests). If it's failsafe, then you can add `-Dit.test=MyTestClass` (see
+the [docs](https://maven.apache.org/surefire/maven-failsafe-plugin/examples/single-test.html) for
+more). You may need to add `-DfailIfNoTests=false` if you end up building more than one module.
+
+> NOTE: run the expected `mvn` command locally before pushing, just to make sure the command runs
+> the test as expected.
+
+Then as before, commit, push and run the CI workflow manually - check its execution to ensure it
+behaves as expected. Try to reproduce the failure with this; if it works, great, that's a much
+faster feedback loop. If it doesn't, then you unfortunately have to revert your changes.
+
+##### Debugging
+
+> NOTE: the end process is somewhat convoluted, so do not hesitate to suggest things which may
+> simplify it.
+
+Next is the fun part where it comes to debugging. You can use the `maven.surefire.debug` (
+or `maven.failsafe.debug` for failsafe) property to attach a debugger to your test processes. By
+default, this will start an agent at port 5005, and wait for a debugger to connect.
+
+However, for most jobs, we run multiple forks, so you can imagine there would be port collisions!
+Furthermore, we don't always want to wait for a debugger to connect before running test.
+
+To circumvent this, you can specify the agent property through the `maven.surefire.debug`
+or `maven.failsafe.debug` properties. To start a debugger agent without waiting your IDE to connect,
+and using a random port, we can set the property
+to: `-agentlib:jdwp=transport=dt_socket,server=y,suspend=n`.
+
+To enable this for CI, find the job which runs your test, and under it, the step running the tests (
+e.g. the command running `mvn verify` or `mvn test` or the likes). There, you can add the
+property (`maven.debug.surefire` or `maven.debug.failsafe`). For example:
+
+```yaml
+      - name: Maven Test Build
+        run: >
+          mvn -B -T2 --no-snapshot-updates
+          -D forkCount=5
+          -D maven.javadoc.skip=true
+          -D skipUTs -D skipChecks
+          -D failsafe.rerunFailingTestsCount=3 -D flaky.test.reportDir=failsafe-reports
+          -D maven.failsafe.debug=-agentlib:jdwp=transport=dt_socket,server=y,suspend=n
+          -P parallel-tests,extract-flaky-tests
+          -pl '!qa/update-tests'
+          verify
+          | tee "${BUILD_OUTPUT_FILE_PATH}"
+```
+
+When the forks are started, they will print out the port which they use for the debugger, e.g.
+
+```
+[INFO] Listening for transport dt_socket at address: 60989
+[INFO] Listening for transport dt_socket at address: 19382
+[INFO] Listening for transport dt_socket at address: 32783
+[INFO] Listening for transport dt_socket at address: 18293
+[INFO] Listening for transport dt_socket at address: 53782
+```
+
+> This is rather inconvenient, and if someone finds a better way, please share! At the moment, this
+> means you'll have to test each port until you find the right one.
```

Comment:
❓ assuming the dynamically assigned debug port is nowhere accessible to be printed from the catch block of the busy loop?

Reply:
I tried to figure that out but couldn't. I do have one idea which I haven't tried out, but might work:

You know your current PID, so you could check all the ports open for listening on your PID, and probe to find which is the debugger agent (somehow?). But yeah, couldn't find any properties being set with the port or any information about it. I suppose you could add a custom surefire/failsafe logger which greps for this log and sets a property? :smile: Very hacky

The other option is to somehow assign the port based on the fork number in a deterministic way, but I also didn't figure how to do this.